### ISSUE TITLE
[meta] Publish Candidate Recommendation

### ISSUE THREAD

**Issue by anssiko** [w3c/presentation-api#276]
The Candidate Recommendation (CR) publication of the Presentation API is looming in the horizon. By publishing a CR we:

* signal the group has met its requirements satisfactorily for a new standard,
* ask the entire W3C membership to provide feedback,
* formally collect implementation experience to demonstrate that the specification works in practice.

While we're well on our way to satisfy the [CR requirements][1], there are some things to be completed before we can hit this significant milestone, namely:

- [x] Resolve the open issues we can address before the F2F (thanks @tidoust):
  - [x] #162 Evaluate Presentation API against the Media Accessibility User Requirements spec
  - [x] #163 PresentationSession should have stream interfaces!
  - [x] #218 Possibility for a character to be interpreted differently depending on locale
  - [x] #241 bufferedAmount property on PresentationConnection
  - [x] #253 Have the receiving browsing context generate the presentation identifier?
  - [x] #254 When is the first presentation connection created in the receiving browsing context?
  - [x] #269 Only allow one PresentationAvailability per PresentationRequest
  - [x] #273 PresentationConnectionClosedEventInit.message needs to have a default value

- [x] Resolve the [open issues tagged for F2F][2] that likely require F2F discussion:
  - [x] #99 Define behavior of Window and Fullscreen APIs in the presentation browsing context
  - [x] #153 Assess interoperability of Presentation API implementations
  - [x] #255 Is a new Permission type required for presentation display availability?
  - [x] #275 User Data Controls in Web Browsers guidelines
  - [x] #311 Add presentation request URL fallback mechanism
  - [x] #318 Clear service worker's persistent storages in a receiving browsing context
- [x] Provide a _draft_ test suite to be able to say we have "document[ed] how adequate implementation experience will be demonstrated". Please get in touch with the Test Facilitator @louaybassbouss to contribute and help.
  - [x] #266 [meta] Presentation API Testing

With continued active participation we're trending toward hitting CR after the May F2F, before the vacation period.

[1]: https://www.w3.org/2015/Process-20150901/#candidate-rec
[2]: https://github.com/w3c/presentation-api/issues?q=is%3Aissue+is%3Aopen+label%3AF2F
[3]: https://github.com/w3c/presentation-api/issues/276

**Comment by tidoust** [comment]
@anssiko, I agree that #202 (Presentations without communication channel) can be considered a new feature and defered to v2. I updated issue labels accordingly.

Although not explicit in the [CR requirements](https://www.w3.org/2015/Process-20150901/#candidate-rec), publication as CR normally implies that known technical issues have been addressed and more generically that the group has solved all issues that could lead to normative changes in the spec. We also need to demonstrate that the specification has undergone wide review.

As such, I would add at least the following issues to the list of issues to be solved before CR:

- #162 Evaluate Presentation API against the Media Accessibility User Requirements spec
- #163 PresentationSession should have stream interfaces!
- #218 Possibility for a character to be interpreted differently depending on locale
- #241 bufferedAmount property on PresentationConnection
- #253 Have the receiving browsing context generate the presentation identifier?
- #254 When is the first presentation connection created in the receiving browsing context?
- #269 Only allow one PresentationAvailability per PresentationRequest
- #273 PresentationConnectionClosedEventInit.message needs to have a default value

Some of these issues may need to be discussed at the F2F, but most of them simply need some sort of "PR" (Pull Request or Proposed Resolution). Essentially, I'd like to propose the same game as in January: let's kill some issues so that the spec may be published as CR shortly after the F2F!

**Comment by anssiko** [comment]
As suggested by @tidoust, let's apply the proven formula from January to the above issue list.

Anyone is free to pick any issue from the above list, address it by submitting a PR, or by suggesting text in the issue comments to be integrated. The results of this collaborative exercise is to be summed up at the F2F.

**Comment by anssiko** [comment]
I updated the first comment in this issue to link to the known CR blockers, split in two buckets: (1) issues we think can be addressed before the F2F, and (2) issues that require F2F discussion. Please see:

https://github.com/w3c/presentation-api/issues/276#issue-145678322

We're making good progress in addressing the first bucket of issues -- contributions from all group participants are welcome!

**Comment by anssiko** [comment]
Here's the summary of the PROPOSED RESOLUTIONs for the remaining CR blockers we discussed at the F2F.

@mfoltzgoogle I assume you'll update the spec to address the first four issues, and will sync with relevant implementers re #311 feasibility.

@tidoust When should we have a publication ready spec for the CR transition request if we'd like to publish by the end of June, or soon after? I'd like to get the CR out before we enter the seasonally slower summer months.

- #99 Define behavior of Window and Fullscreen APIs in the presentation browsing context
  - F2F minutes: https://www.w3.org/2016/05/24-webscreens-minutes.html#item16
  - [ ] **PROPOSED RESOLUTION**: No change to the Presentation API, feature would be addressed by the CSS WG if a strong use case motivates it. Open another issue about recommending to add "TV" to the User-Agent string.
- #153 Assess interoperability of Presentation API implementations
  - F2F minutes: https://www.w3.org/2016/05/24-webscreens-minutes.html#item15
  - [x] **ACTION**:
     - advance with the CG chartering https://github.com/webscreens/cg-charter/ to incubate protocol-level interop
     - spun off an issue #311 impacting API interop
- #255 Is a new Permission type required for presentation display availability?
  - F2F minutes: https://www.w3.org/2016/05/24-webscreens-minutes.html#item13
  - [ ] **PROPOSED RESOLUTION**: no new permission type required
- #275 User Data Controls in Web Browsers guidelines
  - F2F minutes: https://www.w3.org/2016/05/24-webscreens-minutes.html#item14
  - [ ] **PROPOSED RESOLUTION**: complete the clear site data steps in the spec (add AppCache, Service Workers) and add non-normative note referencing TAG document for more complete list, add note clarifying this is not about defining private mode
- #311 Add presentation request URL fallback mechanism
  - F2F minutes: https://www.w3.org/2016/05/25-webscreens-minutes.html#item03
  - [ ] **PROPOSED RESOLUTION**: Add a fallback mechanism by adding a new constructor on PresentationRequest that takes an Array of DOMString, and by exposing the presented URL in an "url" attribute on PresentationConnection, pending feedback from implementers that this is doable.

**Comment by tidoust** [comment]
@anssiko 

> When should we have a publication ready spec for the CR transition request if we'd like to publish by the end of June, or soon after?

Before the spec may be published as CR, we need to organize a transition call with *the Director* (at least one week) and work with Comm team on the transition announcement (a couple of days). That would suggest **Friday 24 June** for a publication beginning of July, and thus **Monday 13 June** to issue the call for consensus.

**Comment by anssiko** [comment]
Current status: good progress, two CR blockers remain: #275 (have a PR #308) and #311 (no PR to review yet).

We aim to send out a CfC soon, but not today.

**Comment by anssiko** [comment]
[CfC: Publish CR of Presentation API, review by 27 June][1]

[1]: https://lists.w3.org/Archives/Public/public-secondscreen/2016Jun/0071.html

**Comment by anssiko** [comment]
Presentation API CR publication status update:

In the [Call for Consensus][1] sent 14 June we set the following condition for the CR publication:

> Please note the group will resolve the issues #311 (URL fallback mechanism) and #318 (clear service worker's persistent storage) per agreement at the F2F and include these changes in the CR.

This condition is not yet met, thus we may want extend the CfC slightly to ensure appropriate review time.

All - this is a great opportunity to help close these two remaining CR blockers so we can advance the spec to CR. All contributions welcome!

[1]: https://lists.w3.org/Archives/Public/public-secondscreen/2016Jun/0071.html

**Comment by anssiko** [comment]
Status update for the remaining CR blockers:
* [x] #311 fixed with #316 (thanks @mfoltzgoogle)
* [ ] #318 has a [proposed fix][1] (thanks @tomoyukilabs) pending review and integration into the spec

Given #316 had a review time of 14 days with no major concerns, and there's a [proposed fix][1] for #318 (smaller interoperability impact), the initial CfC holds and after landing the fix for #318 we proceed with the original plan to publish CR earliest beginning of July.
 
[1]: https://github.com/w3c/presentation-api/issues/318#issuecomment-228297869

**Comment by anssiko** [comment]
Also the draft test suite (tracked in #266) is coming together just in time for the CR publication (thanks @louaybassbouss et al.).

**Comment by anssiko** [comment]
All the CR blockers have now been addressed. Thank you for your contributions everyone, with special thanks to the editor @mfoltzgoogle who has reflected the group's consensus into the spec language.

I'm happy that (once again) we were able to sprint and get the spec ready for publication in time. We're now ready to advance with the CR publication as planned.

**Comment by mfoltzgoogle** [comment]
@anssiko I filed an issue today (#324) that may impact this timeline.  Do you want to wait for a spec change to land for that first?

**Comment by tidoust** [comment]
@mfoltzgoogle The version of the spec that is to be published as CR is somewhat frozen at this stage. It was reviewed by the Director and publication as Candidate Recommendation is scheduled for tomorrow (Thursday). I would suggest that we go ahead with the CR publication as-is, and address #324 and subsequent issues in the Editor's Draft.

From a process perspective, changing the name of an interface is obviously a normative change which would in theory require another publication as CR. In practice, we can argue that changing the name of such an event interface is benign (I think that most apps will simply listen to the `close` event and won't ever check the class of the event they get), so this change should not delay the progress of the spec along the Recommendation track.

**Comment by anssiko** [comment]
I second @tidoust.

**Comment by mfoltzgoogle** [comment]
Yes, Issue #324 naming cleanup and not a semantic change.  If it shouldn't block CR then let's not derail the process underway.

**Comment by anssiko** [comment]
Candidate Recommendation published:
https://www.w3.org/TR/2016/CR-presentation-api-20160714/
